csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3453 @@
+  // }
+
+  // Make an intersection of the exact and loop map. This will group together
+  // entries in each loop group that are exact with eachother. This provides a
+  // better graph to do promotion and replays.
+
+  // It's tempting to use the intersection of the almost exact and loop, but we
+  // need to model broadcast promotion, and if we have two tensors like:
+  //
+  // T1[i0, b1] = T0[i0]
+  // T2[i0, b1] = T0[i0]
+  //
+  // Then resolution of:
+  // T4 = T1[i0, b1] + T3[i0, i1]
+  // T6 = T2[i0, b1] + T5[i0, i2]
+  //
+  // The almost exact map will map T1's and T2's b1 together, but they're being
+  // resolved to i1 and i2 respectively. So we want to have separate entries so
+  // we can have an easy to process promotion map.
+  //
+  // Loop is a permissive like map, it could have many entries, use the exact
+  // map as the one we iterate on to reduce complexity as it hopefully has
+  // smaller groups and this algorithm scales with the number of groups *
+  // (number of entries in groups ^ 2)
+
+  auto intersection_exact_loop_graph = initializeIdGraph();
+  for (auto exact_group :
+       idGraph(IdMappingMode::EXACT).disjointIdSets().disjointSets()) {
+    auto set_size = exact_group->size();
+    for (auto id0_i : c10::irange(set_size)) {
+      auto id0 = exact_group->vector()[id0_i];
+      for (auto id1_i = id0_i; id1_i < set_size; id1_i++) {
+        auto id1 = exact_group->vector()[id1_i];
+        // id0 and id1 map in the almost exact map, if they also map in the loop
+        // graph, then add the mapping to the inersection
+        if (idGraph(IdMappingMode::LOOP)
+                .disjointIdSets()
+                .strictAreMapped(id0, id1)) {
+          intersection_exact_loop_graph.mapIds(id0, id1);
+        }
+      }
+    }
+  }
+
+  // std::cout << "Intersection exact - loop: " << std::endl;
+  // {
+  //   IdGroups groups;
+  //   for (auto group :
+  //        intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+  //     groups.pushBack(group);
+  //   }
+  //   std::cout << debug_print::idGroupsStringShort(groups) << std::endl;
+  // }
+
+  // Promotion logic is going to be on the intersection of the exact and loop
+  // graph. We will generate a map on the entries of this graph so it's
+  // important to not modify this graph moving forward, as that would invalidate
+  // the map.
+  //
+  // iel stands for Intersection of the Exact and Loop graphs.
+  std::unordered_map<IdGroup, IterDomain*> iel_promotion_map;
+
+  // This should probably work just on terminating inputs, as we shouldn't be
+  // able to modify a broadcast domain between root and rfactor which would be
+  // required to resolve a non input broadcast domain. But for now leaving it as
+  // traversal on all broadcast groups.
+  for (auto iel_group :
+       intersection_exact_loop_graph.disjointIdSets().disjointSets()) {
+    if (!iel_group->front()->isBroadcast()) {
```

Comment:
When a broadcast ID is merged with a non-broadcast ID, it becomes a non-broadcast ID. Doesn't such an ID also need a promotion? 

Or, maybe it doesn't matter as long as root IDs are promoted? Is it actually what the above comment is saying?

Reply:
We're preparing to traverse the IEL graph with replays of promoted broadcast id's. In the IEL graph we're identifying which broadcast groups to make not broadcast.

What I'm saying is if we just take the IdGroups that are input IdGroups (no unique definitions registered), we should probably just be able to promote those then replay the IEL graph with that promotion, and that should be good enough.

So what we're just doing is trying to map broadcast IEL IdGroups to their concretized IEL groups, if that concretization is performed within the loop group. If it's not performed within the loop group then it doesn't need to be replayed as if it was.

Reply:
Since broadcast can't happen between root and rfactor, I believe that's why we should be able to just use the inputs of the IEL graph. I think if it could show up there then maybe we couldn't? Or maybe we still could, I don't know.

For now I'm leaving it as happening on all broadcast groups because once we actually use all of this I'll be more confident changing it.

Reply:
> What I'm saying is if we just take the IdGroups that are input IdGroups (no unique definitions registered), we should probably just be able to promote those then replay the IEL graph with that promotion, and that should be good enough.

Makes sense.